File: apps/net/Ch4_OOP/Models/ParkingLot.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace Ch4_OOP.Models
{
    public class ParkingLot: IEnumerable<string>
    {
        private Dictionary<string, Car> spots;

        public ParkingLot(IEnumerable<string> names)
        {
            spots = names.ToDictionary(n => n, name => (Car)null);
        }

        public static ParkingLot Create(int spotsPerLevel, int levels)
        {
            List<string> names = new List<string>();
            string[] level_names = { "A", "B", "C", "D", "E", "F", "G" };
            foreach(var ln in level_names.Take(levels))
            {
                for (int n = 0; n < spotsPerLevel; n++)
                {
                    names.Add($"{ln}{n+1}");
                }
            }

            return new ParkingLot(names);
        }

        public IEnumerator<string> GetEnumerator()
        {
            foreach (var spot in spots.Keys)
            {
                yield return spot;
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            foreach (var spot in spots.Keys)
            {
                yield return spot;
            }
        }

        public string[] TakenSpots()
        {
            return (
                from spot in spots
                where spot.Value != null
                select $"{spot.Key} has the {spots[spot.Key].ModelName}"
                ).ToArray();
        }

        public void ParkCar(Car car)
        {
            foreach (var kv in spots)
            {
                if (kv.Value == null)
                {
                    spots[kv.Key] = car;
                    break;
                }
            }
        }
    }
}

File: apps/py/ch04_oop/models/parking_lot.py
from typing import List, Dict, Optional

from models.car import Car


class ParkingLot:

    def __init__(self, spot_names: List[str]):
        # self.spots = dict()
        # for n in spot_names:
        #     self.spots[n] = None

        self.spots: Dict[str, Optional[Car]] = {
            n: None
            for n in spot_names
        }

    @staticmethod
    def create(spots_per_level: int, levels: int) -> "ParkingLot":
        names = []
        level_names = ['A', 'B', 'C', 'D', 'E', 'G']
        for ln in level_names[:levels]:
            for n in range(1, spots_per_level+1):
                names.append(f'{ln}{n}')

        return ParkingLot(names)

    def park(self, car: Car):
        for k, v in self.spots.items():
            if v is None:
                self.spots[k] = car
                break

    def __iter__(self):
        for i in self.spots.items():
            yield i










